lib/spack/spack/binary_distribution.py

```diff
@@ -1254,8 +1268,26 @@ def _build_tarball(
     # get the sha256 checksum of the tarball
     checksum = checksum_tarball(tarfile_path)
 
-    # add sha256 checksum to spec.json
+    # push tarball to remote mirror
+    # do this first in order to attempt to get an etag for the uploaded resource
+    web_util.push_to_url(
+        spackfile_path,
+        remote_spackfile_path,
+        keep_original=False,
+        # For S3 buckets, enable the extra checksum SHA256 attribute
+        extra_args={"ChecksumAlgorithm": "SHA256"},
+    )
+    # Try to get the ETag of the uploaded tarball. This must happen after uploading the
+    # tarball since it is not possible to pre-compute the resource etag
+    try:
+        spackfile_etag = web_util.read_etag(remote_spackfile_path)
```

Comment:
this has a race condition

Reply:
Is it? I thought the push was blocking. If not, I will see if we can't get it from the header. The reason I didn't do that to begin with is there is, afaict, no actual standard for how an ETag is represented in a response header, or if it is even required to provide one.

Reply:
So we _ can_ get the ETag from the s3 response...but the current method we use, `upload_file` will not work. It requires using the `put_object` method instead.

I am not sure what the connotations of that are. I will keep looking into it.

Reply:
`put_object` has some issues...but I think they are solvable. I will push some changes after I verify the fix and you can let me know if you can think of a less horrible solution.